// @serenity-js/mocha/lib/mappers/MochaTestMapper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MochaTestMapper = void 0;
const io_1 = require("@serenity-js/core/lib/io");
const model_1 = require("@serenity-js/core/lib/model");
/**
 * @package
 */
class MochaTestMapper {
    detailsOf(test) {
        function fileOf(t) {
            switch (true) {
                case !!t.ctx && !!t.ctx.currentTest && !!t.ctx.currentTest.file:
                    return t.ctx.currentTest.file;
                case !!t.file:
                    return t.file;
                case !!t.parent:
                    return fileOf(t.parent);
                default:
                    throw new Error(`Couldn't determine path of ${t}`);
            }
        }
        return new model_1.ScenarioDetails(new model_1.Name(this.nameOf(test)), new model_1.Category(this.featureNameFor(test)), new io_1.FileSystemLocation(new io_1.Path(fileOf(test))));
    }
    featureNameFor(scenario) {
        var _a;
        const parentTitle = (_a = scenario === null || scenario === void 0 ? void 0 : scenario.parent) === null || _a === void 0 ? void 0 : _a.title;
        return parentTitle !== undefined && parentTitle.trim() !== ''
            ? this.featureNameFor(scenario.parent)
            : scenario.title;
    }
    fullNameOf(scenario) {
        return scenario.parent
            ? `${this.fullNameOf(scenario.parent)} ${scenario.title}`.trim()
            : scenario.title;
    }
    nameOf(scenario) {
        return this.fullNameOf(scenario)
            .slice(this.featureNameFor(scenario).length)
            .trim();
    }
}
exports.MochaTestMapper = MochaTestMapper;
